src/metabase/pulse/render.clj

```diff
@@ -336,23 +344,26 @@
         " of "     [:strong {:style (style {:color color-gray-3})} (format-number col-count)]
         " columns."])]))
 
-(defn- render:table
+(s/defn ^:private render:table :- RenderedPulseCard
   [timezone card {:keys [cols rows] :as data}]
-  [:div
-   (render-table (prep-for-html-rendering timezone cols rows nil nil cols-limit))
-   (render-truncation-warning cols-limit (count cols) rows-limit (count rows))])
+  {:attachments nil
+   :content     [:div
+                 (render-table (prep-for-html-rendering timezone cols rows nil nil cols-limit))
+                 (render-truncation-warning cols-limit (count cols) rows-limit (count rows))]})
 
-(defn- render:bar
+(s/defn ^:private render:bar :- RenderedPulseCard
```

Comment:
Because we don't have a call like

```clojure
(s/set-fn-validation! true)
```

anywhere, we have to add `^:always-validate` metadata to every `s/defn` function declaration so Schema will validate their inputs and/or outputs. I did a lot of measuring and the performance impact is negligible to have it validate everything all the time, so in general I think it makes sense to just always validate everything. Otherwise the schemas here don't do anything unless you set bind `*assert*` or call `s/set-fn-validation!`.

Alternatively we should just put a call to `s/set-fn-validation!` somewhere and we can remove all the `^:always-validate` tags.